crates/compiler/gen_llvm/src/llvm/build.rs

```diff
@@ -2964,136 +2964,61 @@ fn list_literal<'a, 'ctx>(
     let list_length = elems.len();
     let list_length_intval = env.ptr_int().const_int(list_length as _, false);
 
-    // TODO re-enable, currently causes morphic segfaults because it tries to update
-    // constants in-place...
-    // if element_type.is_int_type() {
-    if false {
+    let is_all_constant = elems.iter().all(|e| e.is_literal());
+
+    if is_all_constant {
         let element_type = element_type.into_int_type();
+
         let element_width = layout_interner.stack_size(element_layout);
-        let size = list_length * element_width as usize;
-        let alignment = layout_interner
+
+        let refcount_slot_bytes = layout_interner
             .alignment_bytes(element_layout)
-            .max(env.target.ptr_width() as u32);
+            .max(env.target.ptr_width() as u32) as usize;
 
-        let mut is_all_constant = true;
-        let zero_elements =
-            (env.target.ptr_width() as u8 as f64 / element_width as f64).ceil() as usize;
+        let refcount_slot_elements =
+            (refcount_slot_bytes as f64 / element_width as f64).ceil() as usize;
 
-        // runtime-evaluated elements
-        let mut runtime_evaluated_elements = Vec::with_capacity_in(list_length, env.arena);
+        let data_bytes = list_length * element_width as usize;
 
-        // set up a global that contains all the literal elements of the array
-        // any variables or expressions are represented as `undef`
-        let global = {
-            let mut global_elements = Vec::with_capacity_in(list_length, env.arena);
+        assert!(refcount_slot_elements > 0);
 
-            // Add zero bytes that represent the refcount
-            //
-            // - if all elements are const, then we store the whole list as a constant.
-            //      It then needs a refcount before the first element.
-            // - but if the list is not all constants, then we will just copy the constant values,
-            //      and we do not need that refcount at the start
-            //
-            // In the latter case, we won't store the zeros in the globals
-            // (we slice them off again below)
-            for _ in 0..zero_elements {
-                global_elements.push(element_type.const_zero());
-            }
+        let mut bytes = Vec::with_capacity_in(refcount_slot_elements + data_bytes, env.arena);
```

Comment:
Is this bytes or elements? Can we build up elements, but directly write bytes? I think we need that level of control to properly set the refcount.

Reply:
it's elements

Reply:
pretty sure we can't directly write bytes with the llvm api